rosidl_typesupport_introspection_cpp/resource/msg__type_support.cpp.em

```diff
@@ -45,6 +45,18 @@ header_files = [
 @[    end if]@
 #include "@(header_file)"
 @[end for]@
+
+
+namespace rosidl_typesupport_introspection_cpp
+{
+
+template<>
```

Comment:
This is just a forward decl for a function defined lower down, right? If you could you add a comment to that effect? Is it not already declared in the `.hpp.em` header, and if so could we just include that instead? Why is the forward decl needed?

If we do need the forward declaration here - then I want to point out this expression

```
'::'.join([package_name] + list(interface_path.parents[0].parts))
```

Which is repeated many times in this file, and I think could be factored out into a variable at the top level block for easier reading and refactoring. Maybe factor out

```
message_namespace = '::'.join([package_name] + list(interface_path.parents[0].parts))
message_namespaced_name = '::'.join([message_namespace, message-structure.namespaced_type.name])
```

And that would really simplify some of the expressions farther down, make them more readable

Reply:
We are declaring that a special instance of the template function (for the given template argument) is present, and will be defined later. If we don't do this, we will get a compile error. (Usage of template function before instantiation)

I refactored the rest of the code as suggested.